proposals/fate-operator-proposal.md

```diff
@@ -0,0 +1,199 @@
+
+#### FATECluster
+```
+apiVersion: app.kubefate.net/v1beta1
+kind: FateCluster
+metadata:
+  name: fatecluster-sample
+  namespace: fate-9999
+spec:
+  version: v1.4.0
+  partyId: "9999"
+  proxyPort: "30009"
+  partyList:
+  - partyId: "10000"
+    partyIp: "192.168.1.10"
+    partyPort: "30010"
+  egg:
+replica: 1
+  # Points to KubeFATE created by CRD “Kubefate”
+  kubefate:
+    name: kubefate-sample
+    namespace:  kube-fate
+```
+The FateCluster defines a deployment of FATE on Kubernetes. 
+* version defines the FATE version deployed in Kubernetes;
+* partyId defines the FML party’s ID;
+* proxyPort defines the exposed port for exchanging models and parameters between different parties in a FML training. It will be exposed as a node port;
+* partyList defines the parties in a federation which take part in collaboratively learning;
+* egg is the worker nodes of FATE.
+
+#### FateJob
+```
+apiVersion: app.kubefate.net/v1beta1
+kind: FateJob
+metadata:
+  name: fatejob-sample
+  namespace: fate-9999
+spec:
+  # fateClusterRef points to the FATE cluster resource created by FateCluster
+  fateClusterRef: fatecluster-sample
+  jobConf:
+    pipeline: |-
+      {
+        "components": {
+          "secure_add_example_0": {
+            "module": "SecureAddExample"
+          }
+        }
+      }
+    modulesConfig: |-
+      {
+        "initiator": {
+          "role": "guest",
+          "party_id": 9999
+        },
+        "job_parameters": {
+          "work_mode": 1
+        },
+        "role": {
+          "guest": [
+            9999
+          ],
+          "host": [
+            9999
+          ]
+        },
+        "role_parameters": {
+          "guest": {
+            "secure_add_example_0": {
+              "seed": [
+                123
+              ]
+            }
+          },
+          "host": {
+            "secure_add_example_0": {
+              "seed": [
+                321
+              ]
+            }
+          }
+        },
+        "algorithm_parameters": {
+          "secure_add_example_0": {
+            "partition": 10,
+            "data_num": 1000
+          }
+        }
+      }
+```
+FateJob defines the job sent to FATE cluster:
+* fateClusterRef defines the cluster of FATE deployed on Kubernetes. Its value is resource name of FATE cluster created by CRD “FateCluster”;
+* jobConf defines the details of a FML job. It includes:
+   * pipeline: the workflow pipeline of FATE. In FATE, there are many prebuilt algorithm components (ref: https://github.com/FederatedAI/FATE/tree/master/federatedml and https://github.com/FederatedAI/FATE/tree/master/federatedrec), which can be used to train models. The pipeline defines how data are passed through and processed in the whole training flow;
+   * moduleConf: the detail configuration of each algorithm component, e.g. the optimizers, the batch size etc.
+
+### Controller design
```

Comment:
Have you looked at https://github.com/kubeflow/common? Can we reuse the Kubeflow/common’s JobController interface ?

Reply:
+1 on above comment

@terrytangyuan are there more details about JobController interface and backend implementation design?

Reply:
There isn't much documentation on the backend implementation design yet. The [README.md](https://github.com/kubeflow/common/blob/master/README.md) and the comments in [the interface](https://github.com/kubeflow/common/blob/master/pkg/apis/common/v1/interface.go) contains some information.